src/main/java/org/gaul/s3proxy/junit/rules/S3ProxyRule.java

```diff
@@ -0,0 +1,188 @@
+        }
+
+        public Builder withHttps() {
+            useHttps = true;
+            return this;
+        }
+
+        public Builder ignoreUnknownHeaders() {
+            ignoreUnknownHeaders = true;
+            return this;
+        }
+
+        public S3ProxyRule build() {
+            return new S3ProxyRule(this);
+        }
+    }
+
+    private S3ProxyRule(Builder builder) {
+        accessKey = builder.accessKey;
+        secretKey = builder.secretKey;
+
+        Properties properties = new Properties();
+        try {
+            blobStoreLocation = Files.createTempDirectory("S3ProxyRule")
+                    .toFile();
+            properties.setProperty("jclouds.filesystem.basedir",
+                blobStoreLocation.getCanonicalPath());
+        } catch (IOException e) {
+            throw new RuntimeException("Unable to initialize Blob Store", e);
+        }
+
+        BlobStoreContext context = ContextBuilder.newBuilder(
+                    builder.blobStoreProvider)
+                .credentials(accessKey, secretKey)
+                .overrides(properties).build(BlobStoreContext.class);
+
+        S3Proxy.Builder s3ProxyBuilder = S3Proxy.builder()
+            .blobStore(context.getBlobStore())
+            .awsAuthentication(builder.authType, accessKey, secretKey)
+            .ignoreUnknownHeaders(builder.ignoreUnknownHeaders);
+
+        if (builder.secretStorePath != null ||
+                builder.secretStorePassword != null) {
+            s3ProxyBuilder.keyStore(builder.secretStorePath,
+                builder.secretStorePassword);
+        }
+
+        int port = builder.port < 0 ? 0 : builder.port;
+        if (builder.useHttps) {
+            endpointFormat = "https://%s:%d";
+            String endpoint = String.format(endpointFormat, LOCALHOST, port);
+            s3ProxyBuilder.secureEndpoint(URI.create(endpoint));
+        } else {
+            endpointFormat = "http://%s:%d";
+            String endpoint = String.format(endpointFormat, LOCALHOST, port);
+            s3ProxyBuilder.endpoint(URI.create(endpoint));
+        }
+
+        s3Proxy = s3ProxyBuilder.build();
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    protected void before() throws Throwable {
+        logger.debug("S3 proxy is starting");
+        s3Proxy.start();
+        while (!s3Proxy.getState().equals(AbstractLifeCycle.STARTED)) {
+            Thread.sleep(10);
+        }
+        logger.debug("S3 proxy is running");
+    }
+
+    @Override
+    protected void after() {
+        logger.debug("S3 proxy is stopping");
+        try {
+            s3Proxy.stop();
+        } catch (Exception e) {
+            throw new RuntimeException("Unable to stop S3 proxy", e);
+        }
+        logger.debug("S3 proxy has stopped");
+    }
+
+    public URI getUri() {
+        return URI.create(String.format(endpointFormat, LOCALHOST,
+            s3Proxy.getPort()));
+    }
```

Comment:
I'm not convinced this works for HTTPS since S3Proxy also has `getSecureEndpoint`.

Reply:
getSecureEndpoint doesn't exist either, we could add it and then here use a conditional on "useHttps" to determine which URI to return.